File: javascript/robot-name/robot-name.js
// configuration
const padAlpha = 2;
const padDigit = 3;

const asciis = {
    keys: "0123456789abcdefghijklmnop",
    vals: "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
};

const digits = {
    keys: "0123456789",
    vals: "0123456789"
};

const max = asciis.keys.length ** padAlpha * digits.keys.length ** padDigit;

// generate all possible ids
var ids = [];
for (let i = 0; i < max; i++) {
    ids.push(i);
}
// then shuffle
function shuffle(a) {
    var j, x, i;
    for (i = a.length - 1; i > 0; i--) {
        j = Math.floor(Math.random() * (i + 1));
        x = a[i];
        a[i] = a[j];
        a[j] = x;
    }
    return a;
}
var shuffledIds = shuffle(ids);

// python like divmod
function divmod(num, div) {
    return { quotient: Math.floor(num / div), remainder: num % div };
}

// convert decimal to int with custom base and symbols
function decToWhatever(dec, pad, lut) {
    let conv = "";
    for (c of parseInt(dec, 10).toString(lut.keys.length)) {
        conv += lut.vals[lut.keys.indexOf(c)];
    }
    return conv.padStart(pad, lut.vals[0]);
}

// convert decimal to custom format AADDD
function numToName(num) {
    let parts = divmod(num, 10 ** padDigit);
    let alphaPart = decToWhatever(parts.quotient, padAlpha, asciis);
    let digitPart = decToWhatever(parts.remainder, padDigit, digits);
    return alphaPart + digitPart;
}

class Robot {
    constructor() {
        this.setName();
    }

    setName() {
        // pick next avalaible id
        let next = shuffledIds.pop();
        if (next === undefined) {
            throw new Error(
                "Le trop de robots est l'ennemi du mieux de robots"
            );
        }
        this.name = numToName(next);
    }

    getName() {
        return this.name;
    }

    reset() {
        this.setName();
    }
}

module.exports = Robot;

if (require.main === module) {
    const max = 26 ** 2 * 10 ** 3;
    let usedNames = [];
    for (let i = 0; i < max; i++) {
        var newRobot = new Robot();
        usedNames[newRobot.name] = true;
    }
}

File: python/robot-name/robot_name.py
import random

# robot name format config AADDD
ASCIIS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
PADASCII = 2
DIGITS = "0123456789"
PADDIGIT = 3

# max robots
MAX = (len(ASCIIS) ** PADASCII * len(DIGITS) ** PADDIGIT)

# Lame PseudoRandom Generator (LPG)
INIT = random.randint(0, MAX)  # start at random position
CURRENT = INIT
# random step
primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37,
          41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]
r = [i for i in range(1, 100) if (MAX % (MAX // i))
     in primes]

STEP = MAX // random.choice(r)


def nextId():
    global CURRENT
    value = CURRENT
    CURRENT = (CURRENT + STEP) % MAX
    return value


def baseN(num, numerals):  # decimal num into custom base integer
    return ((num == 0) and numerals[0]) or (baseN(num // len(numerals), numerals).lstrip(numerals[0]) + numerals[num % len(numerals)])


def paddedBaseN(num, pad, numerals):  # right padded custom integer
    return str(baseN(num, numerals)).rjust(pad, numerals[0])


class Robot(object):
    def __init__(self):
        self.setName()

    def setName(self):
        # format AADDD
        alpha, digit = divmod(nextId(), 10 ** PADDIGIT)
        self.name = paddedBaseN(alpha, PADASCII, ASCIIS) + \
            paddedBaseN(digit, PADDIGIT, DIGITS)

    def reset(self):
        self.setName()


def main():
    print(paddedBaseN(30, 2, "ABCD"))
    # check if all robots can be created
    # total = len(set([nextId() for _ in range(MAX)]))
    # assert(
    #     total == MAX), f":O collision happened {total} / {MAX}"
    # # randomness sample
    # print([Robot().name for _ in range(100)])


if __name__ == '__main__':
    main()
